### DISCUSSION TITLE
posts/delegating-handler

### DISCUSSION THREAD

**Discussion by TechWatching** [TechWatching/techwatching.dev#13] [Announcements]
Written on 28/10/2019 14:02:44 

URL: https://techwatching.dev/posts/delegating-handler

**Comment by TechWatching** [comment]
Comment written by **Gerald Maale** on **29/05/2021 10:37:48**

<p>Hi,<br>Don't you think using IMemoryCache stores the token in the entire server where notwithstanding who logs in, the token is not actually representing the new user logging in. You consider revising this.</p>

**Reply by TechWatching** [reply]
Comment written by Alex Pavlenko on 07/07/2021 02:46:29

If this is the inner service that should be used for entire application, why not?

**Reply by TechWatching** [reply]
Comment written by Alexandre on 22/08/2021 22:53:54

Good remark but in my sample the call to the API is made as a Service, not as a User: in that case caching the token for the whole server is what I want. But I agree with you,in other cases involving multiple users the implementation should be modified a bit.

**Comment by mohammadshahabrafiq** [comment]
Hello Alexandre,

First of all, thank you for the article. Very informative. As I was reading it I noticed there maybe a race condition in the code.

So in the `UserApiAuthenticationService` class the body of conditional `if (!_memoryCache.TryGetValue("Token", out string token))` could potentially be executed by more than one thread at the same time. If multiple threads hit the this same line of code at same time then you will get multiple cache misses. 

See [here](https://blog.novanet.no/asp-net-core-memory-cache-is-get-or-create-thread-safe/) for more details.

This might be the desired behaviour which is fine but just wanted to make the readers aware of this.

**Reply by TechWatching** [reply]
I don't think it is really a problem for the use case in my sample but that is good to know for other use cases. I did not know about that so thanks for the link and for making the other readers aware. 👍

**Comment by aghazadehm** [comment]
Hi,
If the service gets token in body !!! what I do ? 
Payload: { "Token": "token", "Data": "..." }

**Reply by TechWatching** [reply]
You can also you http message handlers and modify the body of the request to add the token
